File: tools/cs2_weapons.py
#!/usr/bin/env python3
"""CS2's gameplay numbers for the three guns -> AnimationData/cs2_weapons.json.

Straight out of `01_weapon_data/firearm_blocks/weapon_<gun>.vdata`. Fields whose
CS meaning is a formula rather than a value are converted here, and the two the
mod cannot take literally are flagged:

  spread     CS carries `m_flSpread` plus a state-dependent `m_flInaccuracy*`,
             both dimensionless direction perturbations. The mod fires one cone,
             so the standing first shot is atan(spread + inaccuracyStand) in
             degrees - a reading of the units, not of CS's bullet code.
  recoil     `m_flRecoilMagnitude` is in CS's own recoil units and its per-shot
             pattern comes from a seeded generator this port does not have. Only
             the RATIO between guns is used; the absolute scale stays the value
             the mod already had for the AK. Marked estimated until a wall-spray
             measurement settles it.

Two-element arrays in the vdata are [primary, alternate]: unsilenced/silenced for
the M4A1-S, unscoped/scoped for the AWP.

Usage:  python3 tools/cs2_weapons.py [--out FILE]
"""

from __future__ import annotations

import argparse
import json
import math
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
DATA = ROOT / "src/ScCsgoKnives/AnimationData"
VDATA = (Path.home() / "workspaces/CSMCReverse/local_cs2_analysis/all_weapons"
         / "01_weapon_data/firearm_blocks")

# mod name -> the vdata block. All 35 CS2 ships, so a gun's numbers are ready before
# the port reaches it; the mod only draws the ones GunSpec lists.
GUNS = {
    "ak47": "weapon_ak47", "m4a1s": "weapon_m4a1_silencer", "awp": "weapon_awp",
    "aug": "weapon_aug", "bizon": "weapon_bizon", "cz75a": "weapon_cz75a",
    "deagle": "weapon_deagle", "elite": "weapon_elite", "famas": "weapon_famas",
    "fiveseven": "weapon_fiveseven", "g3sg1": "weapon_g3sg1", "galilar": "weapon_galilar",
    "glock18": "weapon_glock", "hkp2000": "weapon_hkp2000", "m249": "weapon_m249",
    "m4a4": "weapon_m4a1", "mac10": "weapon_mac10", "mag7": "weapon_mag7",
    "mp5sd": "weapon_mp5sd", "mp7": "weapon_mp7", "mp9": "weapon_mp9",
    "negev": "weapon_negev", "nova": "weapon_nova", "p250": "weapon_p250",
    "p90": "weapon_p90", "revolver": "weapon_revolver", "sawedoff": "weapon_sawedoff",
    "scar20": "weapon_scar20", "sg556": "weapon_sg556", "ssg08": "weapon_ssg08",
    "taser": "weapon_taser", "tec9": "weapon_tec9", "ump45": "weapon_ump45",
    "usp_silencer": "weapon_usp_silencer", "xm1014": "weapon_xm1014",
}

SCALAR = ["m_nDamage", "m_iMaxClip1", "m_nPrimaryReserveAmmoMax", "m_flCycleTime",
          "m_flRange", "m_flRangeModifier", "m_flArmorRatio", "m_flHeadshotMultiplier",
          "m_flPenetration", "m_flRecoveryTimeStand", "m_flRecoveryTimeCrouch",
          "m_flRecoveryTimeStandFinal", "m_nRecoilSeed", "m_nZoomFOV1", "m_nZoomFOV2",
          "m_nZoomLevels", "m_flDeployDuration", "m_flKillAward",
          "m_flZoomTime0", "m_flZoomTime1", "m_flZoomTime2",
          "m_nNumBullets", "m_flCycleTimeWhenInBurstMode", "m_flTimeBetweenBurstShots"]
PAIRS = ["m_flSpread", "m_flInaccuracyStand", "m_flInaccuracyCrouch", "m_flInaccuracyMove",
         "m_flInaccuracyFire", "m_flInaccuracyJump", "m_flRecoilAngle",
         "m_flRecoilAngleVariance", "m_flRecoilMagnitude", "m_flRecoilMagnitudeVariance",
         "m_flMaxSpeed"]

# The AK's existing hand-fitted camera kick, kept as the anchor for the absolute
# recoil scale so only the ratios between guns change. ESTIMATED.
AK_KICK_PITCH_DEGREES = 1.6


def prefab_text(name: str) -> str:
    """One `<name> = { ... }` block out of the source weapons.vdata."""
    source = VDATA.parent / "source/weapons.vdata"
    if not source.exists():
        return ""
    whole = source.read_text("utf-8", "replace")
    start = whole.find("\n%s = " % name)
    if start < 0:
        start = whole.find("%s = " % name)
        if start < 0:
            return ""
    depth = 0
    for i in range(start, len(whole)):
        if whole[i] == "{":
            depth += 1
        elif whole[i] == "}":
            depth -= 1
            if depth == 0:
                return whole[start:i + 1]
    return ""


def read(stem: str) -> dict:
    text = (VDATA / (stem + ".vdata")).read_text("utf-8", "replace")
    # Five guns - Five-SeveN, Glock-18, P250, Revolver, Taser - leave m_flCycleTime and
    # sometimes more to their `_base` prefab, so the prefab is appended and the firearm
    # block searched first. Every regex below takes the first match.
    base = re.search(r'_base\s*=\s*"([^"]+)"', text)
    if base:
        text = text + "\n" + prefab_text(base.group(1))
    out = {}
    for name in SCALAR:
        # Some of these are written as a pair - the Glock's cycle time is
        # [ 0.15, 0.3 ], primary then burst-mode secondary - so a bare number is tried
        # first and the first element of a list second. Five guns need it:
        # Five-SeveN, Glock-18, P250, Revolver and the Taser.
        m = re.search(r"%s\s*=\s*([\d.\-]+)" % re.escape(name), text)
        if m:
            out[name] = float(m.group(1))
            continue
        m = re.search(r"%s\s*=\s*\[\s*([\d.\-]+)" % re.escape(name), text)
        if m:
            out[name] = float(m.group(1))
            out[name + "_isPair"] = True
    for name in PAIRS:
        m = re.search(r"%s\s*=\s*\[([^\]]*)\]" % re.escape(name), text)
        if m:
            out[name] = [float(x) for x in m.group(1).split(",")]
    m = re.search(r"m_bIsFullAuto\s*=\s*(true|false)", text)
    out["m_bIsFullAuto"] = m.group(1) == "true" if m else None
    m = re.search(r"m_bHideViewModelWhenZoomed\s*=\s*(true|false)", text)
    out["m_bHideViewModelWhenZoomed"] = m.group(1) == "true" if m else None
    m = re.search(r"m_bHasBurstMode\s*=\s*(true|false)", text)
    out["m_bHasBurstMode"] = m.group(1) == "true" if m else None
    m = re.search(r'm_eSilencerType\s*=\s*"(\w+)"', text)
    out["m_eSilencerType"] = m.group(1) if m else None
    # The sound events CS2 plays. SPECIAL1 is the silenced shot on the two detachable
    # silencers and on the integrated MP5-SD, where it repeats the unsilenced entry.
    block = re.search(r"m_aShootSounds\s*=\s*\{(.*?)\n\t*\}", text, re.S)
    sounds = {}
    if block:
        for key, value in re.findall(r'(WEAPON_SOUND_\w+)\s*=\s*soundevent:"([^"]+)"', block.group(1)):
            sounds.setdefault(key, value)
    out["m_aShootSounds"] = sounds
    return out


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--out", type=Path, default=DATA / "cs2_weapons.json")
    args = ap.parse_args()

    raw = {gun: read(stem) for gun, stem in GUNS.items()}
    ak_magnitude = raw["ak47"]["m_flRecoilMagnitude"][0]
    kick_per_unit = AK_KICK_PITCH_DEGREES / ak_magnitude

    guns = {}
    for gun, v in raw.items():
        # The Taser is not a firearm and CS2 gives it no spread, inaccuracy, recoil or
        # move speed. Those read as zero rather than being invented, and the derived
        # degrees below come out zero with them.
        pair = lambda k: v.get(k) or [0.0, 0.0]
        spread, alt_spread = pair("m_flSpread")[0], pair("m_flSpread")[-1]
        stand, alt_stand = pair("m_flInaccuracyStand")[0], pair("m_flInaccuracyStand")[-1]
        magnitude, alt_magnitude = pair("m_flRecoilMagnitude")[0], pair("m_flRecoilMagnitude")[-1]
        guns[gun] = {
            "Damage": v["m_nDamage"],
            "HeadshotMultiplier": v["m_flHeadshotMultiplier"],
            "ArmorRatio": v["m_flArmorRatio"],
            "Penetration": v["m_flPenetration"],
            "Magazine": int(v["m_iMaxClip1"]),
            "ReserveClips": int(v.get("m_nPrimaryReserveAmmoMax", 0)),
            "CycleSeconds": v["m_flCycleTime"],
            "FullAuto": v["m_bIsFullAuto"],
            "RangeUnits": v["m_flRange"],
            "RangeModifier": v["m_flRangeModifier"],
            "MaxSpeed": v.get("m_flMaxSpeed") or [0.0],
            # Pellets per shot: 1 for everything but the shotguns (Nova 9, MAG-7 and
            # Sawed-Off 8, XM1014 6).
            "Pellets": int(v.get("m_nNumBullets") or 1),
            # Burst mode exists on exactly two guns. CS2 carries the burst's cycle time
            # and the gap between its shots but not how many shots it fires; three is
            # Counter-Strike's long-standing burst and is marked as the assumption.
            "HasBurstMode": bool(v.get("m_bHasBurstMode")),
            "BurstCycleSeconds": v.get("m_flCycleTimeWhenInBurstMode"),
            "BurstShotSeconds": v.get("m_flTimeBetweenBurstShots"),
            "BurstShotsAssumed": 3 if v.get("m_bHasBurstMode") else 0,
            # WEAPONSILENCER_NONE / _DETACHABLE (M4A1-S, USP-S) / _INTEGRATED (MP5-SD).
            "SilencerType": v.get("m_eSilencerType"),
            "ShootSounds": v.get("m_aShootSounds") or {},
            "ZoomFov": [v.get("m_nZoomFOV1"), v.get("m_nZoomFOV2")],
            "ZoomLevels": int(v.get("m_nZoomLevels", 0)),
            # How long CS2 takes to interpolate to each zoomed FOV. The AWP's is 0.05
            # for all three, three frames at 60 fps, which is why its scope reads as
            # instant. 0.17.1 gated the lens overlay on a 0.25 s aim blend while the
            # world FOV changed immediately, so a quarter second of the shot was
            # magnified with no scope drawn.
            "ZoomSeconds": [v.get("m_flZoomTime0"), v.get("m_flZoomTime1"), v.get("m_flZoomTime2")],
            "HideViewModelWhenZoomed": v.get("m_bHideViewModelWhenZoomed"),
            "RecoveryTimeStand": v.get("m_flRecoveryTimeStand", 0.0),
            "RecoilSeed": v.get("m_nRecoilSeed"),
            "RecoilAngleVariance": pair("m_flRecoilAngleVariance"),
            "RecoilMagnitude": pair("m_flRecoilMagnitude"),
            # Converted, with the reasoning in this file's docstring.
            "SpreadDegrees": round(math.degrees(math.atan(spread + stand)), 5),
            "SpreadDegreesAlternate": round(math.degrees(math.atan(alt_spread + alt_stand)), 5),
            "MoveSpreadDegrees": round(math.degrees(math.atan(spread + pair("m_flInaccuracyMove")[0])), 5),
            "KickPitchDegrees": round(magnitude * kick_per_unit, 5),
            "KickPitchDegreesAlternate": round(alt_magnitude * kick_per_unit, 5),
            "KickYawDegrees": round(magnitude * kick_per_unit
                                    * math.sin(math.radians(pair("m_flRecoilAngleVariance")[0] / 2)), 5),
            "KickRecoverPerSecond": round(1.0 / v["m_flRecoveryTimeStand"], 5) if v.get("m_flRecoveryTimeStand") else 0.0,
            "Raw": v,
        }
        g = guns[gun]
        print("%-6s dmg %g x%g hs, armor %g, mag %d, cycle %gs, falloff %g/500u"
              % (gun, g["Damage"], g["HeadshotMultiplier"], g["ArmorRatio"],
                 g["Magazine"], g["CycleSeconds"], g["RangeModifier"]))
        print("       spread %.4f deg standing (%.4f alt), %.3f deg moving; kick %.3f deg (%.3f alt), "
              "yaw +-%.3f, recover %.2f/s"
              % (g["SpreadDegrees"], g["SpreadDegreesAlternate"], g["MoveSpreadDegrees"],
                 g["KickPitchDegrees"], g["KickPitchDegreesAlternate"], g["KickYawDegrees"],
                 g["KickRecoverPerSecond"]))

    doc = {
        "Format": "ScCsgoKnives.Cs2Weapons/1",
        "Notes": ("Values from weapon_<gun>.vdata. Damage falls off as "
                  "damage * RangeModifier^(units/500), the CS convention documented at "
                  "counterstrike.fandom.com/wiki/Damage_dropoff and "
                  "steamcommunity.com/sharedfiles/filedetails/?id=2599082552 - community "
                  "documentation of the formula, not the SDK source. Spread is "
                  "atan(m_flSpread + m_flInaccuracy*) in degrees, a reading of the units. "
                  "Recoil ratios are CS2's; the absolute degrees-per-recoil-unit scale is "
                  "anchored to the AK's previously hand-fitted 1.6 deg and is ESTIMATED."),
        "UnitsPerMetre": 1.0 / 0.0254,
        "FalloffUnits": 500.0,
        "KickDegreesPerRecoilUnit": round(kick_per_unit, 6),
        "Guns": guns,
    }
    args.out.write_text(json.dumps(doc, indent=1), "utf-8")
    print("wrote %s (%.1f KB)" % (args.out.name, args.out.stat().st_size / 1024))


if __name__ == "__main__":
    main()

File: src/ScCsgoKnives/Animation/Cs2Weapons.cs
using System.IO;
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Serialization;
using Engine;

namespace Game;

/// <summary>
/// CS2's gameplay numbers, from AnimationData/cs2_weapons.json (tools/cs2_weapons.py,
/// out of weapon_&lt;gun&gt;.vdata).
///
/// Damage in Survivalcraft attack power is CS damage unchanged: a Survivalcraft
/// player has 1.0 health and CS damage 36 costs 0.36 of it, the same fraction a
/// 36-damage hit costs a 100-health CS player. Falloff follows CS's
/// damage * RangeModifier^(units/500); the "/500" is community documentation of the
/// formula (counterstrike.fandom.com/wiki/Damage_dropoff and the Steam guide
/// 2599082552 agree), not the SDK source, and is recorded as such.
///
/// Two numbers here are not literal readings, and cs2_weapons.json says so:
/// spread is atan(m_flSpread + m_flInaccuracy*) - a reading of what those
/// dimensionless perturbations mean - and the recoil scale is anchored to the AK's
/// previously hand-fitted 1.6 degrees, so only the RATIOS between guns are CS2's.
/// </summary>
public static class Cs2Weapons {
    public sealed class Gun {
        [JsonPropertyName("Damage")]
        public float Damage { get; set; }
        [JsonPropertyName("HeadshotMultiplier")]
        public float HeadshotMultiplier { get; set; }
        [JsonPropertyName("ArmorRatio")]
        public float ArmorRatio { get; set; }
        [JsonPropertyName("Penetration")]
        public float Penetration { get; set; }
        [JsonPropertyName("Magazine")]
        public int Magazine { get; set; }
        [JsonPropertyName("ReserveClips")]
        public int ReserveClips { get; set; }
        [JsonPropertyName("CycleSeconds")]
        public float CycleSeconds { get; set; }
        [JsonPropertyName("FullAuto")]
        public bool FullAuto { get; set; }
        [JsonPropertyName("RangeUnits")]
        public float RangeUnits { get; set; }
        [JsonPropertyName("RangeModifier")]
        public float RangeModifier { get; set; }
        [JsonPropertyName("MaxSpeed")]
        public float[] MaxSpeed { get; set; }
        /// <summary>Pellets per shot: 1 for everything but the shotguns (Nova 9, MAG-7 and Sawed-Off 8, XM1014 6).</summary>
        [JsonPropertyName("Pellets")]
        public int Pellets { get; set; } = 1;
        /// <summary>Only the Glock-18 and the FAMAS have one.</summary>
        [JsonPropertyName("HasBurstMode")]
        public bool HasBurstMode { get; set; }
        [JsonPropertyName("BurstCycleSeconds")]
        public float? BurstCycleSeconds { get; set; }
        [JsonPropertyName("BurstShotSeconds")]
        public float? BurstShotSeconds { get; set; }
        /// <summary>
        /// CS2 carries the burst's cycle time and the gap between its shots but not how
        /// many it fires. Three is Counter-Strike's burst and the generator writes it as
        /// BurstShotsAssumed, named so the estimate is visible in the data.
        /// </summary>
        [JsonPropertyName("BurstShotsAssumed")]
        public int BurstShotsAssumed { get; set; }
        /// <summary>WEAPONSILENCER_NONE, _DETACHABLE (M4A1-S, USP-S) or _INTEGRATED (MP5-SD).</summary>
        [JsonPropertyName("SilencerType")]
        public string SilencerType { get; set; }
        /// <summary>CS2's own sound events, by WEAPON_SOUND_* key.</summary>
        [JsonPropertyName("ShootSounds")]
        public Dictionary<string, string> ShootSounds { get; set; }
        [JsonPropertyName("ZoomFov")]
        public float?[] ZoomFov { get; set; }

        public bool SilencerDetachable => SilencerType == "WEAPONSILENCER_DETACHABLE";
        public bool SilencerIntegrated => SilencerType == "WEAPONSILENCER_INTEGRATED";
        /// <summary>The event CS2 plays for a shot, silenced or not; null when unknown.</summary>
        public string ShootSound(bool silenced) {
            if (ShootSounds is null) return null;
            if (silenced && ShootSounds.TryGetValue("WEAPON_SOUND_SPECIAL1", out string special)) return special;
            return ShootSounds.TryGetValue("WEAPON_SOUND_SINGLE", out string single) ? single : null;
        }
        [JsonPropertyName("ZoomLevels")]
        public int ZoomLevels { get; set; }
        /// <summary>
        /// CS2's m_flZoomTime per level: how long it interpolates to each zoomed FOV.
        /// The AWP's are all 0.05, three frames at 60 fps, which is why its scope
        /// reads as instant. 0.17.1 held the lens overlay back for a 0.25 s aim blend
        /// while the world FOV changed on the key frame.
        /// </summary>
        [JsonPropertyName("ZoomSeconds")]
        public float?[] ZoomSeconds { get; set; }
        [JsonPropertyName("HideViewModelWhenZoomed")]
        public bool HideViewModelWhenZoomed { get; set; }
        [JsonPropertyName("RecoveryTimeStand")]
        public float RecoveryTimeStand { get; set; }
        [JsonPropertyName("SpreadDegrees")]
        public float SpreadDegrees { get; set; }
        [JsonPropertyName("SpreadDegreesAlternate")]
        public float SpreadDegreesAlternate { get; set; }
        [JsonPropertyName("MoveSpreadDegrees")]
        public float MoveSpreadDegrees { get; set; }
        [JsonPropertyName("KickPitchDegrees")]
        public float KickPitchDegrees { get; set; }
        [JsonPropertyName("KickPitchDegreesAlternate")]
        public float KickPitchDegreesAlternate { get; set; }
        [JsonPropertyName("KickYawDegrees")]
        public float KickYawDegrees { get; set; }
        [JsonPropertyName("KickRecoverPerSecond")]
        public float KickRecoverPerSecond { get; set; }
    }

    sealed class WeaponsFile {
        [JsonPropertyName("Format")]
        public string Format { get; set; }
        [JsonPropertyName("UnitsPerMetre")]
        public float UnitsPerMetre { get; set; }
        [JsonPropertyName("FalloffUnits")]
        public float FalloffUnits { get; set; }
        [JsonPropertyName("Guns")]
        public Dictionary<string, Gun> Guns { get; set; }
    }

    const string Resource = "AnimationData.cs2_weapons.json";
    const string ExpectedFormat = "ScCsgoKnives.Cs2Weapons/1";

    /// <summary>Null when the file loaded; the reason otherwise. See Cs2Effects.LoadError.</summary>
    public static string LoadError { get; private set; } = "not loaded";

    static readonly WeaponsFile s_file = Load();

    /// <summary>Gameplay numbers follow GunNumbers, not GunProfile: the look and the feel switch separately.</summary>
    public static bool Active => KnifeTuning.GunNumbers >= 0.5f && s_file?.Guns is { Count: > 0 };

    public static Gun Get(string gun) =>
        gun is not null && s_file?.Guns is not null && s_file.Guns.TryGetValue(gun, out Gun g) ? g : null;

    /// <summary>
    /// Damage after CS's range falloff, for a shot that travelled `metres`.
    /// Returns the fallback unchanged when the cs2 profile is off or the gun is unknown.
    /// </summary>
    public static float DamageAt(string gun, float metres, float fallback) {
        Gun g = Active ? Get(gun) : null;
        if (g is null || g.RangeModifier <= 0f || g.RangeModifier >= 1f) return g?.Damage ?? fallback;
        float units = MathUtils.Max(metres, 0f) * s_file.UnitsPerMetre;
        return g.Damage * MathF.Pow(g.RangeModifier, units / s_file.FalloffUnits);
    }

    /// <summary>
    /// Cone half-angle in degrees for this shot. CS raises inaccuracy continuously with
    /// speed rather than at a threshold, so the standing and moving values from the
    /// vdata are interpolated by the player's speed as a fraction of the gun's own
    /// m_flMaxSpeed. Both endpoints are read; the linear blend between them is this
    /// port's choice, since CS's own curve is in code the export does not carry.
    /// </summary>
    public static float SpreadDegrees(string gun, bool alternate, float metresPerSecond, float fallback) {
        Gun g = Active ? Get(gun) : null;
        if (g is null) return fallback;
        float standing = alternate ? g.SpreadDegreesAlternate : g.SpreadDegrees;
        float max = g.MaxSpeed is { Length: > 0 } ? g.MaxSpeed[0] / s_file.UnitsPerMetre : 0f;
        if (max <= 0.01f) return standing;
        float f = MathUtils.Saturate(MathUtils.Max(metresPerSecond, 0f) / max);
        return MathUtils.Lerp(standing, g.MoveSpreadDegrees, f);
    }

    public static (float Pitch, float Yaw, float Recover) Kick(string gun, bool alternate,
        float pitch, float yaw, float recover) {
        Gun g = Active ? Get(gun) : null;
        if (g is null) return (pitch, yaw, recover);
        return (alternate ? g.KickPitchDegreesAlternate : g.KickPitchDegrees,
                g.KickYawDegrees, g.KickRecoverPerSecond);
    }

    static WeaponsFile Load() {
        try {
            Assembly assembly = typeof(Cs2Weapons).Assembly;
            string name = assembly.GetManifestResourceNames().FirstOrDefault(n => n.EndsWith(Resource, StringComparison.OrdinalIgnoreCase));
            if (name is null) {
                LoadError = $"no embedded {Resource}";
                KnifeDiagnostics.WarnOnce("cs2-weapons-missing", $"No embedded {Resource}.");
                return null;
            }
            using Stream stream = assembly.GetManifestResourceStream(name);
            // Pinned by JsonPropertyName; PropertyNameCaseInsensitive would not have
            // caught a separator change, which is how Cs2Effects broke in 0.16.4.
            WeaponsFile file = JsonSerializer.Deserialize<WeaponsFile>(stream);
            if (file?.Format != ExpectedFormat || file.Guns is null) {
                LoadError = $"{Resource} is not {ExpectedFormat}";
                KnifeDiagnostics.WarnOnce("cs2-weapons-format", $"{Resource} is not {ExpectedFormat}.");
                return null;
            }
            KnifeLog.Information("[ScCsgoKnives] CS2 weapon data: " + string.Join("; ", file.Guns.Select(kv =>
                $"{kv.Key} dmg={kv.Value.Damage:0.#} falloff={kv.Value.RangeModifier:0.##}/500u "
                + $"spread={kv.Value.SpreadDegrees:0.###}deg kick={kv.Value.KickPitchDegrees:0.###}deg")));
            LoadError = null;
            return file;
        }
        catch (Exception e) {
            LoadError = $"{e.GetType().Name}: {e.Message}";
            KnifeDiagnostics.WarnOnce("cs2-weapons-load", $"Could not read {Resource}: {e.Message}");
            return null;
        }
    }
}
